### ISSUE TITLE
kubernetes secret list permission

### ISSUE THREAD

**Issue by mattharrigan** [containous/traefik#7097]
This is a follow up to [this](https://github.com/containous/traefik/issues/7092) issue. It is a bug, and a major security vulnerability, to give traefik access to read all secrets via the list RBAC. Requiring traefik to have secret list permissions is not acceptable. List allows "for collections, including full object content" per the [docs](https://kubernetes.io/docs/reference/access-authn-authz/authorization/#determine-the-request-verb). This is not a question.  It is a blocker for using the traefik kubernetes provider. I really like traefik and look forward to using it once this is resolved. Thank you.

**Comment by dtomcej** [comment]
Hello @mattharrigan,

You are correct when you state that granting list RBAC allows for access to the content.

We initialize the informer here:

https://github.com/containous/traefik/blob/master/pkg/provider/kubernetes/crd/client.go#L172

and you will notice that it is namespaced. If you would like to restrict secrets access to a specific namespace, you can do so using the namespaces (https://docs.traefik.io/v2.2/providers/kubernetes-crd/#namespaces) filter. This will allow traefik to list and watch only the namespaces you provide.

Go applications that use the Go-client have no ability to tell what RBAC permissions they have. This is demonstrated by your observation in the first ticket: Granting list on a secret allowed the informer to create a lister, but did not allow the associated watcher to be created, nor the associated Get that usually accompanies the lister.

The issue is tied back to the initialization of the informer: By default, Traefik will attempt to watch all secrets, so that if you create a new namespace, and create an ingressRoute, and reference a new secret, Traefik will be able to read it.

If you are more concerned about security and require less flexibility and dynamic discovery, you can restrict it via namespacing, but currently there is no implementation of dynamic watchers for individual secrets, it is done via namespace.

**Comment by mattharrigan** [comment]
> nor the associated Get that usually accompanies the lister

I disagree. The `list` permission can read the secret contents without `get` permissions. This is the core of the security vulnerability. Listing the names of the secrets is not that huge of a deal. Reading the contents is.

In my cluster there are many other secrets in even the most minimal namespace. Some of them are effectively admin tokens. Allowing read access to them is not acceptable.

**Comment by dtomcej** [comment]
@mattharrigan, I think you misunderstood that statement.

The `get` in the go-client bypasses the informer cache, and queries the API directly. List is effectively a superset of get. This is why if you grant list, it is usually assumed that you are also granting get (since list can read all contents anyways).

In Traefik's case, to accomplish what you are asking, would require keeping track of every secret referenced, and then creating informers/listers and event handlers dynamically and keeping them synchronized, while ensuring that we don't duplicate anything.

This possibility was raised in #1784.

I don't disagree with the proposal of this feature, I'm just saying that this is not a bug, but a feature proposal.

**Comment by mattharrigan** [comment]
I think its a bug, not a feature request, since traefik k8s provider currently _requires_ being configured in a fundamentally insecure manner. However debating that point is probably not helpful. I would simply like it resolved.

There are probably many possible solutions. I am not a familiar with the details or implementation of traefik, but in my use case there are only a small number of secrets for certs, so one option could be to not give traefik any permissions for secrets and instead provide them by file. Its all the traefik CRD's that are highly dynamic, and I really like how traefik gets/watches them directly via the built in k8s client.

wrt my previous quote, I don't think traefik ever tries to use `get` directly. It appears to use the [Lister](https://github.com/containous/traefik/blob/master/pkg/provider/kubernetes/crd/client.go#L340). So by not giving `get`, it doesn't really prevent anything from being created because the code doesn't try to create them anyways. But again I'm coming in cold to this codebase.

**Comment by mattharrigan** [comment]
I think we agree that its ok to give traefik read access to secrets it needs. Where we disagree is that I think namespace granularity for RBAC is at best fragile and commonly impossible. k8s has a rich and rapidly evolving ecosystem. Its not reasonable to assume nothing else is operating in the namespace other than traefik. Those other things may rely on the assumption that a `secret` is "secret".

To throw out another possibility, create a new CRD for the cert secrets in the traefik api space. Granting list, get, watch to that resource would be acceptable IMO.

**Comment by mattharrigan** [comment]
This blog [post](https://www.cyberark.com/resources/threat-research-blog/securing-kubernetes-clusters-by-eliminating-risky-permissions) says that listing secrets is the number 1 example of a risky permission.

**Comment by mattharrigan** [comment]
I would very much appreciate this issue being resolved. One specific feature request would be that if traefik doesn't have secret list permissions, log that fact, but do not exit with an error. This is very much analogous to phone apps which request may permissions for location, contacts, etc, but in case they are not granted the app continues to operate, albeit with reduced capability. My hope is that this more specific feature request is relatively easy and non controversial. Thank you.

**Comment by Binyang2014** [comment]
From my point of view, there may be two use cases for traefik in k8s.
1. certs are stored in a **single** namespace and a centralized auth server does the authentication and authorization check. In this case. We  only need to access limited namespace to get required certs and the namespace name can be known before deploy traefik. (But backend service and ingress object maybe created in any namespace dynamically)
2. Each service store crerts in their own namespace and  dose auth separately. Since service can be deployed dynamically, we don't know which namespace should be watched when deploying traefik.

For case 1, I think we can add a startup option for traefik, which can point out secrets in which namespace should be listed. Then we can init the informer cache correctly. 
For case 2, we can just keep current behaviors.

@dtomcej Is that make sense?

**Comment by jpetazzo** [comment]
It looks like Traefik 2.5 runs fine if it doesn't have permission to access Secrets. (Of course, in that case, TLS keys and certs can't be stored in Secrets.)

It would be nice if Traefik could use TLS secrets if it only has `get` permission to these secrets (as opposed to `list`), but as far as I understand, this would require to change a lot of code, because instead of setting up a single informer for all secrets, Traefik would need to set up one separate `watch` connection for each secret referenced in an Ingress object. Beyond the significant refactoring that this would require, I don't even know if that's a desirable behavior (what would be the load on the API server if we have one watch connection per referenced secret?). I honestly don't know what's the best behavior here.

**Comment by Binyang2014** [comment]
You can correct, I revisit the ingress spec and find it can only reference the secret in the same namespace. It's not a easy work.

BTW: remove secrets list permission from cluster role don't work for me. The treafik in my cluster failed to route requests.   :-<

**Comment by jnonino** [comment]
Hi @rtribotte, I understand your point, but I still have the question about why Traefik needs to include secrets in the cluster resource discovery. A K8S cluster might have lots of secrets unrelated to Traefik behaviour and does not look correct for Traefik to require access to all those secrets. And the situation is worst when you find out that `list` permissions to all secrets allows to actually read the value of the secret.

And regarding your last paragraph,
>Even if Traefik would not rely on informers and not require the list permission, it would still require the get permission in a whole namespace, which in a way is pretty much the same as list permission, by allowing to read any secret (as long as the requested name is indeed matching).

I don't see why Traefik would still permissions to read any secrets. If you know which is the requested name and do a match to get it and read, you can ask the users to have that only secret available for Traefik with that specific name. And request permissions only for that secret.

**Comment by rtribotte** [comment]
Hello @jnonino,

To sum up the situation regarding the required RBAC verbs for secrets resource with Traefik:

| RBAC verbs for secrets    | Discovery | Events (secret update) | Errors |
| -------- | ------- | ------- | ------- |
| list, watch  | OK    | OK | - |
| list, get  | OK    | NOK | ```E0227 08:39:45.501950       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: Failed to watch *v1.Secret: unknown (get secrets)``` |
| get, watch  | NOK    | NOK | ```W0227 08:38:00.439389       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik-ingress-controller" cannot list resource "secrets" in API group "" at the cluster scope```<br>```E0227 08:38:00.439449       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: Failed to watch *v1.Secret: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik-ingress-controller" cannot list resource "secrets" in API group "" at the cluster scope``` |
| get  | NOK    | NOK |```W0227 08:41:36.897369       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik-ingress-controller" cannot list resource "secrets" in API group "" at the cluster scope```<br>```E0227 08:41:36.897411       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: Failed to watch *v1.Secret: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik-ingress-controller" cannot list resource "secrets" in API group "" at the cluster scope``` |
| list  | OK    | NOK | ```E0227 08:39:45.501950       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: Failed to watch *v1.Secret: unknown (get secrets)``` |
| watch  | NOK    | NOK | ```W0227 08:38:00.439389       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik-ingress-controller" cannot list resource "secrets" in API group "" at the cluster scope```<br>```E0227 08:38:00.439449       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.4/tools/cache/reflector.go:229: Failed to watch *v1.Secret: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik-ingress-controller" cannot list resource "secrets" in API group "" at the cluster scope``` |

We give more thoughts to this issue, and we have a proposal.
We could introduce a [field selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/) option that would be used for the secrets informers at initialization (e.g.:  `metadata.name=supersecret`). This selector would allow the discovery of secrets based on their names, and also based on the namespace, depending if using a `Role` or a `ClusterRole` for Traefik service.

This field selector should be then used in combination with narrowed RBACs, by using the `resourceNames` option:
```yaml
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: traefik-ingress-controller
  namespace: default

rules:
  - apiGroups:
      - ""
    resources:
      - secrets
    verbs:
      - get
      - list
      - watch
    resourceNames:
      - supersecret
...
```

This would allow to guarantee that verbs `list` and `watch`, needed by the usage of informers, is limited only to a very specific subset of secrets in the cluster.

WDYT?

**Comment by jnonino** [comment]
Hi @rtribotte, I think that ti would be get if Traefik stops requiring a wide access to all secrets in a namespace or the whole cluster.
I actually got into this discussion with a PR in traefik-helm-chart adding what you are suggesting at the end of your  comment: https://github.com/traefik/traefik-helm-chart/pull/1009.

Do you think this use of field selectors can be included in next Traefik release? We are running Traefik v2.11.0 installed via the official Helm chart you provide so we are ready to update and adopt this new feature once it is released!!!

**Comment by rtribotte** [comment]
Hello @jnonino,

This feature makes a lot of sense. Unfortunately, this would not make it to our roadmap for a while as we are focused elsewhere. So we cannot commit this to be in the next Traefik release (which will not be a v2.x, as any enhancement will now land on v3.x). If you or another community member would like to build it, let us know, and we will work with you to ensure you have all the information needed to merge it.

Don't forget to check out the [contributor docs](https://github.com/traefik/contributors-guide/blob/master/pr_guidelines) and link the PR to this issue.

**Comment by justinrush** [comment]
This issue seems to have carried over to the gateway API provider. If I use the helm chart setting to restrict access to specific secrets the gateway provider never starts (the gateway resource is listed as waiting on provider) and the pod logs errors:

```
W0612 14:28:14.816188       1 reflector.go:539] k8s.io/client-go@v0.29.2/tools/cache/reflector.go:229: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik" cannot list resource "secrets" in API group "" at the cluster scope
E0612 14:28:14.816221       1 reflector.go:147] k8s.io/client-go@v0.29.2/tools/cache/reflector.go:229: Failed to watch *v1.Secret: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:default:traefik" cannot list resource "secrets" in API group "" at the cluster scope
```

Like the original poster, this is a hard blocker for us. We have been exploring migrating from nginx chroot to traefik with the Gateway API. Given that your focus is on the gateway API, is this something that can be fixed in that provider?

**Comment by justinrush** [comment]
@rtribotte is the proposal to add some cluster label selector attribute on the `pkg/provider/kubernetes/crd/clientWrapper` so that it can use the `NewSharedInformerFactoryWithOptions` with the label selector filter rather than the `NewSharedInformerFactory` when setting the `factoryClusterScope`?

I haven't dug through how that bubbles up in your config, but if thats the proposed solution I'd be open to taking a look at it. Any information or background you have would be helpful.

**Comment by rtribotte** [comment]
Hello @justinrush,

Sorry for the late reply.

I think it would rather be a field selector configuration for the `WithTweakListOptions` to use with the Kubernetes Secret informers factory:
https://github.com/traefik/traefik/blob/b1b4e6b918e8eeaf9e24823baf24dbc77f7d373e/pkg/provider/kubernetes/crd/client.go#L229

**Comment by justinrush** [comment]
> Hello @justinrush,
> 
> Sorry for the late reply.
> 
> I think it would rather be a field selector configuration for the `WithTweakListOptions` to use with the Kubernetes Secret informers factory:
> 
> https://github.com/traefik/traefik/blob/b1b4e6b918e8eeaf9e24823baf24dbc77f7d373e/pkg/provider/kubernetes/crd/client.go#L229

Makes sense to me. I'll work on setting up a local dev/test environment and submit a PR.

**Comment by justinrush** [comment]
ok, PR https://github.com/traefik/traefik/pull/10894 submitted, @mattharrigan / @jnonino FYI in case you have thoughts on ways this should exposed in the helm chart.
